로그인 로직 개발

sjhjack opened the pull request:
## 📑 개요

<!-- 개요에는 왜 이 작업을 했는지 알려주세요! -->

- #41 

<!-- 예시는 다음과 같습니다. -->
<!-- #(해당 이슈 번호) -->
<!-- 중복 회원 가입 방지 기능 구현 -->
<!-- Assignees 에는 자신과 참여를 원 하시는 분을 선택하시면 됩니다! -->

### ✅ PR 체크리스트

---

- [X] 🔀 PR 제목의 형식을 잘 작성했나요?
- [X] 💯 테스트는 잘 통과했나요?
- [X] 🏗️ 빌드는 성공했나요?
- [X] 🧹 불필요한 코드는 제거했나요?
- [X] 💭 이슈는 등록했나요?
- [X] 🏷️ 라벨은 등록했나요?

### 🚀 상세 작업 내용

---

<!-- 상세 작업 내용 에서는 어떠한 작업을 했는지 상세히 적어주세요! -->

<!-- 예시는 다음과 같습니다. -->
<!-- - [X] 아이디 중복 검사 비즈니스 로직 구현 -->
<!-- - [X] 중복일 경우 예외 처리 기능 구현 -->

- [x] UserInfo 요청 Google API 추가
- [x] Birthday 요청 Google API 추가
  - 생일 비공개 유저 저장 로직 추가
- [x] GoogleOAuthClient 모듈의 public 메서드에 설명 주석 추가
- [x] AuthController, AuthService 작성
- [x] 로그인/회원가입 기능 구현
  - 로그인/회원가입 분기 처리
    - DB 조회 O : 로그인 처리 (토큰 정보 저장)
    - DB 조회 X : 회원가입 처리 (회원 정보, 토큰 정보 저장)
- [x] 토큰 재발급 기능 구현
  - 토큰 유효성 검증 분기 처리
    - Redis 조회 O : 토큰 재발급 에러 (401)
    - Redis 조회 X : 잘못된 토큰 에러 (400)
- [x] 회원 탈퇴 기능 구현
  - Redis에서 토큰 정보 삭제
  - 회원 정보 Hard Delete -> 추후 논의 후 변경 예정
- [x] Redis Token 모듈 작성
  - {key : value} = {access_token : refresh_token} 으로 저장 (refresh_token 만료기한 만큼)
  - 토큰 정보 저장, 삭제, 조회 기능 작성
- [x] 전체 테스트 코드 작성

### 📁 ETC

---

closed: #41 

<!-- 이 곳에서는 관련 자료나 사진을 올여주세요! -->
<!-- 링크를 넣고 싶은 경우에는 MAC 에서는 커맨드 + K, Windows 에서는 컨트롤 + K를 누르면 [](url) 가 생성되는데 [] 안에는 원하시는 링크의 제목을 입력하고 () 안에는 URL을
입력해주세요! -->
<!-- 사진 같은 경우에는 drag and drop 으로 사진을 추가할 수 있습니다! -->

sjhjack committed:
Move(#41): AuthMemberServiceImpl 위치 변경

sjhjack committed:
Fix(#41): SocialMember 쿼리 생성 오류 해결

sjhjack committed:
Refactor(#41): AuthMemberServiceImplTest의 SocialMember 생성 코드 수정

sjhjack committed:
Move(#41): SocialMemberRepository 위치 변경

sjhjack committed:
Feat(#41): Google UserInfo 요청 API 추가

sjhjack committed:
Refactor(#41): SocialMemberRepository import 위치 변경

sjhjack committed:
Feat(#41): SocialMemberRepositoryUtils 추가

sjhjack committed:
Refactor(#41): utils의 static 메서드 적용

sjhjack committed:
Feat(#41): AuthTokensResponse 추가

sjhjack committed:
Feat(#41): GoogleUserInfoResponse 추가

sjhjack committed:
Refactor(#41): Google Token 요청 주석 작성

sjhjack committed:
Refactor(#41): Google Token 재발급 요청 주석 작성

sjhjack committed:
Refactor(#41): Google 유저 정보 요청 주석 작성

sjhjack committed:
Refactor(#41): Google 토큰 유효성 검증 요청 주석 작성

sjhjack committed:
Feat(#41): auth-domain, redis-token-domain 모듈 추가

sjhjack committed:
Feat(#41): AuthService 토큰 재발급 메서드 작성

sjhjack committed:
Feat(#41): Redis 설정 추가

sjhjack committed:
Feat(#41): auth-domain 예외 처리 추가

sjhjack committed:
Refactor(#41): 생성되는 RedisTemplate 이름 변경

sjhjack committed:
Feat(#41): wypl-core에 auth-domain 의존성 추가

sjhjack committed:
Feat(#41): AuthServiceImpl 에 로그인/회원가입 로직 추가

sjhjack committed:
Feat(#41): SocialMemberRepository 에 exits 쿼리 추가

sjhjack committed:
Feat(#41): AuthDomainServiceImpl 에 토큰 존재 여부 확인, 회원가입 데이터 저장, 토큰 저장 메서드 추가

sjhjack committed:
Feat(#41): 토큰 저장 로직 추가

sjhjack committed:
Refactor(#41): memberId 로직 모듈화

sjhjack committed:
Feat(#41): 로그인 로직에 토큰 저장 로직 추가

sjhjack committed:
Refactor(#41): not exists -> exists로 변경

sjhjack committed:
Feat(#41): 생일 요청 API의 response 추가

sjhjack committed:
Feat(#41): GoogleOAuthClient에 생일 요청 API 추가

sjhjack committed:
Fix(#41): #48 충돌 해결

sjhjack committed:
Refactor(#41): 생일 response null 체크 메서드 추가

sjhjack committed:
Refactor(#41): 필요 없는 주석 삭제

sjhjack committed:
Refactor(#41): 파라미터 타입 변경

sjhjack committed:
Feat(#41): Redis에서 Refresh Token 값 가져오는 메서드 추가

sjhjack committed:
Feat(#41): 토큰 유효성 검증 시, 토큰 재발급 분기 처리 추기

sjhjack committed:
Refactor(#41): 유저 정보 요청 시, 예외 처리 추가

sjhjack committed:
Refactor(#41): Auth Header 세팅 코드 모듈화

sjhjack committed:
Refactor(#41): 생일 요청 메서드의 response.getBody().emptyBirthday() 부분 개선

sjhjack committed:
Refactor(#41): 생일 정보 요청 메서드의  주석 작성

sjhjack committed:
Feat(#41): Redis Token Exception 추가

sjhjack committed:
Feat(#41): Redis 에서 refresh token 조회 시, null 예외처리 추가

sjhjack committed:
Feat(#41): service 단 토큰 재발행 추가

sjhjack committed:
Feat(#41): Redis에서 토큰 삭제

sjhjack committed:
Feat(#41): service 단 토큰 삭제 기능 추가

sjhjack committed:
Feat(#41): 로그아웃, 회원탈퇴 서비스 로직 추가

sjhjack committed:
Chore(#41): 자동정렬 수정

sjhjack committed:
Feat(#41): 부정문을 private 메서드로 분리

sjhjack committed:
Feat(#41): 로그인 처리 테스트 코드 작성

sjhjack committed:
Refactor(#41): 로그인 처리 테스트 코드를 Nested로 변경

sjhjack committed:
Refactor(#41): 테스트 코드 메서드 명 변경

sjhjack committed:
Feat(#41): 회원가입 테스트 코드 작성

sjhjack committed:
Refactor(#41): 잘못 작성된 코드 삭제

sjhjack committed:
Refactor(#41): 회원 탈퇴 로직 수정

sjhjack committed:
Refactor(#41): 토큰 재발행 요청의 유효성 검증 로직 수정

sjhjack committed:
Feat(#41): 토큰 재발행 테스트 코드 작성

sjhjack committed:
Feat(#41): 로그아웃 테스트 코드 작성

sjhjack committed:
Feat(#41): 회원탈퇴 테스트 코드 작성

sjhjack committed:
Refactor(#41): TokenRepository 관련 코드 마이그레이션

sjhjack committed:
Refactor(#41): test 의존성을 최상단으로 변경

sjhjack committed:
Feat(#41): AuthDomainServiceImpl 테스트 코드 및 MemberFixture 추가

sjhjack committed:
Refactor(#41): @Transactional 위치를 Repository에서 Service로 변경

sjhjack committed:
Refactor(#41): 토큰 만료 시간 저장 코드 간소화

sjhjack committed:
Refactor(#41): Redis 테스트를 위해 EmbeddedRedis 설정 변경

sjhjack committed:
Refactor(#41): deleteToken 로직 변경

sjhjack committed:
Feat(#41): TokenRepository 테스트 코드 작성

sjhjack committed:
Refactor(#41): 테스트 코드에 display name 추가

sjhjack committed:
Feat(#41): Member Repository의 save 테스트 코드 작성

sjhjack committed:
Feat(#41): SocialMember Repository의 save 테스트 코드 작성

sjhjack committed:
Refactor(#41): 중복 멤버 저장 테스트 코드 리펙토링

sjhjack committed:
Refactor(#41): MemberRepositoryTest에 @DataJpaTest 적용

sjhjack committed:
Feat(#41): SocialMember 중복 저장 테스트 추가

sjhjack committed:
Feat(#41): AuthController 작성

sjhjack committed:
Chore(#41): import 순서 변경

sjhjack committed:
Chore(#41): restdocs 의존성 추가

sjhjack committed:
Feat(#41): AuthControllerTest 작성

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,52 @@
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
```
사용하지 않는 기능의 어노테이션은 삭제해주시면 감사합니다!

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/auth/service/AuthServiceImpl.java:
```diff
@@ -0,0 +1,115 @@
+			throw new GoogleOAuthException(GoogleOAuthErrorCode.NOT_AUTHORIZATION_MEMBER);
+		}
+
+		// FIXME : member Id를 같이 보내야할까? 안 보내도 될 것 같은데.. redis에 access_token : {refresh_token, member_id}로 저장?
```
음 이 내용에 대해서는 `memberId`를 전송하지 않아도 괜찮을 것 같습니다!

sjhjack replied:
같이 고민해주셔서 감사합니다!

KIMSEI1124 commented on application/wypl-core/src/test/java/com/wypl/wyplcore/auth/controller/AuthControllerTest.java:
해당 테스트가 Github Actions에서 실패하는 것으로 파악됩니다! 원인을 분석해보시고 알려주세요!

sjhjack replied:
원인 알아보고 공유하겠습니다!

KIMSEI1124 commented on domain/redis-token-domain/src/test/java/com/wypl/redistokendomain/TokenRepositoryTest.java:
```diff
@@ -0,0 +1,113 @@
+			boolean result = tokenRepository.checkExistsToken(accessToken);
+
+			// Then
+			assertThat(result).isEqualTo(false);
```
`assertThat(result).isEqualTo(false);` 대신에 `assertThat(result).isFalse();` 가 있습니다! 해당 내용을 사용해는건 어떤가요?

sjhjack replied:
좋은 지적 감사합니다! 가독성이 향상되는 장점이 있겠네요. 반영하겠습니다.

KIMSEI1124 commented on domain/jpa-member-domain/src/test/java/com/wypl/jpamemberdomain/SocialMemberRepositoryTest.java:
```diff
@@ -0,0 +1,82 @@
+			SocialMember findSocialMember = socialMemberRepository.findById(result.getId()).get();
+
+			// Then
+			assertThat(result.getId()).isEqualTo(findSocialMember.getId());
+			assertThat(result.getMember().getMemberId()).isEqualTo(findSocialMember.getMember().getMemberId());
+			assertThat(result.getOauthId()).isEqualTo(findSocialMember.getOauthId());
+			assertThat(result.getOauthProvider()).isEqualTo(findSocialMember.getOauthProvider());
```
이렇게 하나씩 하는 것 보다 재귀적으로 확인할 수 있는 테스트 방법이 있습니다!

해당 방법으로 수정해보시고, 발생하는 문제점을 해결해보시면 좋을 것 같습니다!

sjhjack replied:
확인해보니 `isEqualTo()` 메서드는 Object끼리 비교가 가능하여 `assertThat(result).isEqualTo(findSocialMember)` 으로 수정했습니다!
의도하신 방법이 맞을까요?

KIMSEI1124 replied:
다른 방법을 의도하였는데! 해당 방법으로 성공이 되었군요! 아마도 `BaseEntity`를 `extends` 하지 않아서 그런 것 같습니다!

```java
assertThat(foundMember)
				.usingRecursiveComparison()
				.ignoringFields("created", "lastModified")
				.isEqualTo(savedMember);
```

다음과 같이 객체 내부의 필드 값들에 대해서 재귀적으로 검증하면서 일부 검증이 불필요한 내용을 제외하는 방법이 있습니다! 추후 다른 테스트를 진행하실 때 사용해 보면 좋을 것 같아요!

sjhjack replied:
이런 방법도 있었군요! 배워갑니다. 감사합니다 😄

KIMSEI1124 commented on domain/jpa-member-domain/src/main/java/com/wypl/jpamemberdomain/member/utils/SocialMemberRepositoryUtils.java:
```diff
@@ -0,0 +1,14 @@
+import com.wypl.jpamemberdomain.member.repository.SocialMemberRepository;
+
+public class SocialMemberRepositoryUtils {
+	public static SocialMember getSocialMember(SocialMemberRepository socialMemberRepository, String id) {
+		return socialMemberRepository.findByOauthProviderAndOauthId(OauthProvider.GOOGLE, id)
+			.orElseThrow(() -> new MemberException(MemberErrorCode.NO_SUCH_MEMBER));
+	}
```
해당 내용을 보아하니 구글에 종속이 되어 보입니다!

만약 다른 로그인 방식이 생긴다면 어떻게 처리하실건가요?!

sjhjack replied:
좋은 지적 감사합니다!
현재 생각으로는 `OauthProvider`를 파라미터로 전달받도록 할 것 같습니다.
메서드를 사용하는 모든 곳에서 수정해야 한다는 번거로움이 있지만, 아직 다른 방법이 떠오르지 않습니다 😢 

혹시 고려해볼 방법이 있을까요?

KIMSEI1124 replied:
사실 파라미터를 제외하면 생각할 방법이 메서드의 이름에다 어떤 제공자에게 인증을 받았는지 적어주는 방법이 생각납니다!

```java
public static SocialMember getSocialMemberByGoogle(SocialMemberRepository socialMemberRepository, String id) {
    return socialMemberRepository.findByOauthProviderAndOauthId(OauthProvider.GOOGLE, id)
	.orElseThrow(() -> new MemberException(MemberErrorCode.NO_SUCH_MEMBER));
}
```

하지만 새로운 제공자가 생기면 메서드가 하나씩 늘어난다는 단점이 있습니다! 한번 고민해 보시고 선택하면 좋을 것 같아요!

KIMSEI1124 commented on domain/jpa-member-domain/src/main/java/com/wypl/jpamemberdomain/member/data/MemberDto.java:
해당 Dto는 어떠한 이유때문에 생성한 Dto인가요?!

이름이 모호해서 다른 Dto가 생긴다면 헷갈릴 수 있다고 생각합니다!

sjhjack replied:
회원가입 로직에서 저장할 데이터를 담기 위한 Dto 입니다.
의도를 명확히 하기 위해 `MemberSaveDto` 처럼 이름을 변경하는게 낫다고 생각하시나요?

KIMSEI1124 replied:
저는 개인적으로 클래스의 이름을 명확히 작성해주는게 더 좋다고 생각합니다! 😄

KIMSEI1124 commented on domain/jpa-member-domain/src/main/java/com/wypl/jpamemberdomain/member/data/SocialMemberDto.java:
`MemberDto` 와 동일한 질문입니다!

sjhjack replied:
`MemberDto`와 동일합니다!

KIMSEI1124 commented on domain/auth-domain/src/test/java/com/wypl/authdomain/auth/service/fixture/MemberFixture.java:
```diff
@@ -0,0 +1,55 @@
+import com.wypl.jpamemberdomain.member.domain.Member;
+import com.wypl.jpamemberdomain.member.domain.SocialMember;
+
+// Todo : Access Level 고민해보자
+// @AllArgsConstructor
```
주석 내용을 해결해주시면 감사합니다!

sjhjack replied:
`@AllArgsConstructor`가 인식되지 않는 원인을 아직 해결하지 못해서 해당 부분 주석 처리 후 생성자를 명시적으로 추가해놓은 상태입니다.
해결하는 대로 더 고민해보겠습니다!

KIMSEI1124 commented on domain/auth-domain/src/main/java/com/wypl/authdomain/auth/service/AuthDomainServiceImpl.java:
```java
@Transactional(readOnly = true)
@RequiredArgsConstructor
@Service
public class AuthServiceImpl {
```

에서는 `@Transactional(readOnly = true)`를 작성하셨는데 해당 서비스에서는 안하신 이유가 있을까요?!

sjhjack replied:
제대로 확인하지 못한 것 같습니다!
해당 클래스를 다시 살펴보니, `saveAuthData`와 `deleteToken`에도 `@Transactional`이 누락된 것을 확인했습니다.
누락된 내용도 포함해서 수정하겠습니다.

KIMSEI1124 reviewed (changes requested):
해당 이슈 처리하시느라 고생하셨습니다! 👍🏼 

몇 가지가지 궁금한 사항들작성해 두었고었수정해야 하는하는 목록에 대해서도 코멘트를 달아두었습니다!
하시다가 궁금하신 내용이 있다질문해 주시면시면 감사합니다!

sjhjack committed:
Chore(#41): 사용하지 않는 어노테이션 삭제

sjhjack committed:
Chore(#41): 재발행 관련 고민내용 주석 삭제

sjhjack committed:
Refactor(#41): false 검증 메서드 변경

isEqualTo(false) -> isFalse()

sjhjack committed:
Refactor(#41): 누락된 @Transactional 추가 및 readOnly 속성 추가

sjhjack committed:
Refactor(#41): SocialMember 동등비교 로직 수정

sjhjack committed:
Update backend-server-test.yml

clean test -> test -i로 변경

KIMSEI1124 commented on .github/workflows/backend-server-test.yml:
```diff
@@ -27,7 +27,7 @@ jobs:
 
       - name: 🧪 Spring Boot Test
-        run: ./gradlew clean test
+        run: ./gradlew test -i
```
음.. 해당 github actions 의 명령어를 `./gradlew clean test -i`로 해도 좋을 것 같아요!

sjhjack replied:
실행 속도는 느려지지만 최신 상태에서 테스트 하는 것이 더 중요해보이네요. 반영하겠습니다!

sjhjack committed:
Chore(#41): application.yml에 test 프로필 추가

sjhjack committed:
Chore(#41): main의 application.yml에서 test 프로필 삭제

sjhjack committed:
Feat(#41): test 용 yml 파일 추가

sjhjack committed:
Refactor(#41): AuthControllerTest에 WebMvcTest 적용

sjhjack committed:
Chore(#41): test 용 application.yml 주석 처리

sjhjack committed:
Feat(#41): test 용 application 추가

KIMSEI1124 commented on application/wypl-core/src/main/java/com/wypl/wyplcore/auth/service/AuthServiceImpl.java:
```diff
@@ -0,0 +1,113 @@
+
+	@Transactional
+	public void quitMember(AuthMember authMember) {
+		// Todo : 회원 탈퇴 로직 논의
```
해당 논의가 필요하다는 내용으로 이슈를 하나 만들어주세요! 👍🏼

sjhjack replied:
정리해서 올려보겠습니다!

KIMSEI1124 commented on application/wypl-core/src/test/java/com/wypl/wyplcore/auth/controller/AuthControllerTest.java:
```diff
@@ -0,0 +1,172 @@
+@AutoConfigureRestDocs
+// @AutoConfigureMockMvc
+// @SpringBootTest
+@WebMvcTest(AuthController.class)
```
Good! 따로 작성하신 문서를 봤는데 잘 해결하신 것 같아요!

그렇다면 `@ControllerTest` 와 같은 어노테이션을 만들어서 처리할 수 있을 것 같은데 따로 이슈를 만들어서 해결해보는건 어떨까요?!

sjhjack replied:
좋은 의견입니다!
두 어노테이션을 합해서 만들어보겠습니다.

KIMSEI1124 commented on application/wypl-core/src/test/resources/application.yml:
해결되었다면 해당 내용은 필요 없을 것 같습니다!

만약 주석으로 처리한다면 🔥 삭제하는게 좋을 것 같아요!

sjhjack replied:
넵! 문제가 해결되었으니 필요 없는 부분은 모두 삭제하겠습니다.

KIMSEI1124 reviewed (approved):
전체적으로 잘 된 것 같습니다!!

해당 이슈 개발하느라 엄청 고생많으셨습니다! 👍🏼 
먼저 `Approve`를 해두었습니다! 몇몇 희망사항을 적어봤는데 한번 봐주시면 좋을 것 같습니다!

sjhjack committed:
Remove(#41): 깃허브 에러 해결 후 필요 없는 파일 제거

sjhjack committed:
Rename(#41): Dto 클래스 이름 명확히 변경

Dto의 의도가 명확하지 않다는 피드백을 반영하여 아래와 같이 변경했습니다.
- MemberDto -> MemberSaveDto
- SocialMemberDto -> SocialMemberSaveDto

sjhjack committed:
Refactor(#41): git workflows 테스트 명령어에 clean 추가

실행 속도는 느려지지만 최신 상태에서 테스트 하는 것이 더 중요하다고 생각하여 clean 명령어 추가

sjhjack committed:
Chore(#41): 주석 제거

sjhjack committed:
Refactor(#41): @ContextConfiguration 추가

sjhjack committed:
Chore(#41): test 용 더미 파일 추가
